### ISSUE TITLE
[Feature request] Create a lua API to allow data transmission through FSO channel

### ISSUE THREAD

**Issue by krevett62** [scp-fs2open/fs2open.github.com#1301]
As discussed in this [url=http://www.hard-light.net/forums/index.php?topic=93312.new#new]topic[/url] it would be a nice feature to be able to communicate between host and client through lua scripting in FSO, allowing some scripts in multiplayer. An additional request would be to be able to hook players individually through scripting while playing a MP game of FSO (a kind of #hv.Player hook). Of course this will be for future releases as 3.8.0 is in a code freeze state ;)
Thanks!

**Comment by asarium** [comment]
The Lua side of this request should be relatively easy to handle. It would require one function for sending data to one or all players and one callback for receiving data. To make sure that scripts don't conflict with each other it would be a good idea to introduce some sort of message namespace which could be assigned by the script. The script would then use that id to send its messages and it could register a callback for when a message with that ID arrives.

I don't know enough about the multi code to determine how easy the implementation of this would be. I looked at the SEXP multi code and it looks like it would be relatively easy to add another packet type which contains the serialized Lua data. So far, this is how I would image the API would look like:
```lua
net.isMaster() -- true if we run the game
net.isClient() -- true if we are just a normal client
net.Players -- A list of all players in the game. I'm not sure if this list is valid on a client
net.Master -- Handle to the master player of the game

-- receiver would either be nil or one of the handles from net.Players. A client may only send data to net.Master. message_id would be a unique string that identifies what script is sending this data. data would be the lua value to send. This would be restricted to nil, numbers, booleans, strings and tables of these types (the other types can't be serialized)
net.sendMessage(receiver, message_id, data)

-- This registers a function to be called when a message with the specified id has been received. The passed function gets a handle to the player that sent this message and the data itself
net.registerReceiver(message_id, function(sender, data) end)
```

As far as I can tell, the current system does not allow to send data from one client to another. This is not a restriction since this can be emulated by routing the message via the server which would then send the message to the right player. This would need to be implemented by a script if required.

**Comment by JohnAFernandez** [comment]
I am currently looking into doing something like this, although it may not get done in time for the multi bump.  I actually thought of a problem with it: you would want both parties to have the same version of the script automatically.

**Comment by asarium** [comment]
I don't think that is an issue that FSO on its own can fix since we don't really know what script belongs to which mod. However, this is similar to a bigger issue that is inconsistent mod data in general (e.g. Player A has version 1.2 but B has 1.5). That is something the launcher (Knossos) could help us with.

It could compute a hash of all the files related to the launched mod (a hash of all the file hashes actually) and pass that to FSO. FSO could then exchange that hash with the master and that compares if it has a different mod than the player and reject them or show a message.
